### PR TITLE
[ft-template] Introduce Default vs custom mode + Restructure the readme and config files to be more beginner friendly

### PR DESCRIPTION
**Pull request by kouroshHakha**
# What does this PR do? 

1. Adds a new concept for default vs. custom entry mode
2. Updates the readme accordingly to more clearly explain llmforge as a library and its features, etc.
3. Updates the config files and their structure so that they are found more easily, and have better docs.

# Verification process

I reran all of the new configs with the latest llmforge upgrade in this [PR](https://github.com/anyscale/llm-forge/pull/425) (almost ready to be merged) and verified that they run at least for 1 iteration per epoch for 3 epochs and do not oom during checkpointing (RAM and GPU)

Here is the complete grid of experiments: https://docs.google.com/spreadsheets/d/1_bA2UHagTXNMz5PWT2xdAxoJgdzCZkYa8tQCVd1DkyQ/edit?gid=0#gid=0

Takeaway: 
- For some reason the runs on llama-3-70B were insanely slow on g5 (even its lora) I think we need to do some profiling and optimization.

### CODE REVIEW

File: templates/fine-tune-llm_v2/README.ipynb [hunk 1]
```diff
@@ -6,142 +6,341 @@
...
+   "cell_type": "markdown",
    "metadata": {},
-   "outputs": [],
    "source": [
-    "# [Optional] You can set the WandB API key to track model performance\n",
-    "# import os\n",
-    "# os.environ[\"WANDB_API_KEY\"]=\"YOUR_WANDB_API_KEY\"\n",
+    "### Models Supported in Default Mode\n",
+    "\n",
+    "Default mode supports a select list of models, with a fixed cluster type of 8xA100-80G. For each model we only support context lengths of 512 up to Max. context length in increments of 2x (i.e. 512, 1024, ...). Here are the supported models and their configurations:\n",
```
Comment (SumanthRH) [inline templates/fine-tune-llm_v2/README.ipynb:82]: Nit: Can we call this list with a better name to reference everywhere? Like "Core Models" .

File: templates/fine-tune-llm_v2/README.ipynb [hunk 2]
```diff
@@ -6,142 +6,341 @@
...
+    "        {\"role\": \"assistant\", \"content\": \"[[1]]\"},\n",
+    "    ]\n",
+    "}\n",
+    "```\n",
+    "\n",
+    "Prompt Format for llama-3-instruct (YAML):\n",
+    "\n",
+    "```yaml\n",
+    "system: \"<|start_header_id|>system<|end_header_id|>\\n\\n{instruction}<|eot_id|>\"\n",
+    "user: \"<|start_header_id|>user<|end_header_id|>\\n\\n{instruction}<|eot_id|>\"\n",
```
Comment (SumanthRH) [inline templates/fine-tune-llm_v2/README.ipynb:140]: FYI we need a small llmforge change for this prompt format to be supported. `trailing_assistant` is currently a required field.  (i'll take it up soon)

File: templates/fine-tune-llm_v2/README.ipynb [hunk 3]
```diff
@@ -6,142 +6,341 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "---------\n",
+    "**Task:** \n",
+    "\n",
+    "Fine-tune llama-3-8b-instruct in custom mode (model is supported in default mode) on 16xA10s (auto mode uses 8xA100-80G) with context length of 512.\n",
```
Comment (SumanthRH) [inline templates/fine-tune-llm_v2/README.ipynb:277]: Nit: "Fine-tune llama-3-8b-instruct in custom mode (model is supported in default mode)" for a place like this the `model is supported in default mode` phrase is wordy and a bit confusing. A shortname like "core models" or something is useful.

File: templates/fine-tune-llm_v2/README.ipynb [hunk 4]
```diff
@@ -6,142 +6,341 @@
...
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "## Step 1 - Launch a fine-tuning run in [workspaces](https://docs.anyscale.com/platform/workspaces/)\n",
+    "## What is LLMForge?\n",
+    "\n",
+    "LLMForge is a library that implements a collection of design patterns that use Ray, RayTrain, and RayData in combination with other open-source libraries (e.g. Deepspeed, 🤗 Huggingface accelerate, transformers, etc.) to provide an easy to use library for fine-tuning LLMs. In addition to these design patterns, it offers tight integrations with the Anyscale platform, such as model registery, streamlined deployment, observability, Anyscale's job submission, etc."
```
Comment (SumanthRH) [inline templates/fine-tune-llm_v2/README.ipynb:42]:
```suggestion
    "LLMForge is a library that implements a collection of design patterns that use Ray, RayTrain, and RayData in combination with other open-source libraries (e.g. Deepspeed, 🤗 Huggingface accelerate, transformers, etc.) to provide an easy to use library for fine-tuning LLMs. In addition to these design patterns, it offers tight integrations with the Anyscale platform, such as model registry, streamlined deployment, observability, Anyscale's job submission, etc."
```

File: templates/fine-tune-llm_v2/README.ipynb [hunk 5]
```diff
@@ -6,142 +6,341 @@
...
+    "system_in_user: False\n",
+    "```\n",
+    "\n",
+    "#### Use-case: Casual language modeling (aka continued pre-training), custom prompt formats (e.g. Llama-guard):\n",
+    "\n",
+    "Example Continued pre-training (JSON):\n",
+    "```json\n",
+    "{\n",
+    "    \"messages\": [\n",
+    "        {\"role\": \"user\", \"content\": \"Once upon a time ...\"},\n",
```
Comment (SumanthRH) [inline templates/fine-tune-llm_v2/README.ipynb:151]: We need `llmforge` changes to support this :) 

Currently, the minimum requirement for a `messages` entry is 
`[{"role": "user", "content": "n"}, {"role": "assistant", "content": "n"}]` 
(non-empty content, atleast one user and assistant message)


Your choice on sequencing this PR and the llmforge changes, but we need two changes now:
- supporting just one user role (promptformat validation change)
- making `trailing_assistant` optional 

and both can be out in the next release

File: templates/fine-tune-llm_v2/README.ipynb [hunk 6]
```diff
@@ -6,142 +6,341 @@
...
+    "\n",
+    "* Extend the context length of the model via methods like RoPE scaling.\n",
+    "\n",
+    "### ✅ Configurability of hyper-parameters\n",
+    "\n",
+    "* Full control over learning hyper-parameters such as learning rate, n_epochs, batch size, etc.\n",
+    "\n",
+    "### ✅ Anyscale and third-party integrations\n",
+    "\n",
+    "* (Coming soon) Model registery: \n",
```
Comment (SumanthRH) [inline templates/fine-tune-llm_v2/README.ipynb:193]:
```suggestion
    "* (Coming soon) Model registry: \n",
```

File: templates/fine-tune-llm_v2/training_configs/custom/meta-llama/Meta-Llama-3-70B/lora/8xA100-80G-4k.yaml [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+# By default, we will keep the best checkpoint. You can change this to keep more checkpoints.
+num_checkpoints_to_keep: 1
+
+# Deepspeed configuration, you can provide your own deepspeed setup
+deepspeed:
+  config_path: deepspeed_configs/zero_3_offload_optim+param.json
+
+# Accelerator type, we value of 0.001 is not important, as long as it is
+# beteween 0 and 1. This ensures that accelerator type is used per trainer
+# worker.
```
Comment (SumanthRH) [inline templates/fine-tune-llm_v2/training_configs/custom/meta-llama/Meta-Llama-3-70B/lora/8xA100-80G-4k.yaml:39]:
```suggestion
# Accelerator type, the value of 0.001 is not important, as long as it is
# between 0 and 1. This ensures that the given accelerator is available for each trainer
# worker.
```

File: templates/fine-tune-llm_v2/training_configs/custom/meta-llama/Meta-Llama-3-70B/lora/32xA10-4k.yaml [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+# By default, we will keep the best checkpoint. You can change this to keep more checkpoints.
+num_checkpoints_to_keep: 1
+
+# Deepspeed configuration, you can provide your own deepspeed setup
+deepspeed:
+  config_path: deepspeed_configs/zero_3_offload_optim+param.json
+
+# Accelerator type, we value of 0.001 is not important, as long as it is
+# beteween 0 and 1. This ensures that accelerator type is used per trainer
+# worker.
```
Comment (SumanthRH) [inline templates/fine-tune-llm_v2/training_configs/custom/meta-llama/Meta-Llama-3-70B/lora/32xA10-4k.yaml:42]:
```suggestion
# Accelerator type, the value of 0.001 is not important, as long as it is
# between 0 and 1. This ensures that the given accelerator is available for each trainer
# worker.
```

File: templates/fine-tune-llm_v2/training_configs/custom/meta-llama/Meta-Llama-3-70B/full/32xA10-4k.yaml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  config_path: deepspeed_configs/zero_3_offload_optim+param.json
+
+# Rank-0 (aka trainer) should have 140 GB memory
+# This memory is required for weight aggregation.
+trainer_resources:
+  memory: 150_323_855_360 # 140 GB memory
+
+# Accelerator type, we value of 0.001 is not important, as long as it is
+# beteween 0 and 1. This ensures that accelerator type is used per trainer
+# worker.
```
Comment (SumanthRH) [inline templates/fine-tune-llm_v2/training_configs/custom/meta-llama/Meta-Llama-3-70B/full/32xA10-4k.yaml:44]:
```suggestion
# Accelerator type, the value of 0.001 is not important, as long as it is
# between 0 and 1. This ensures that the given accelerator is available for each trainer
# worker.
```

File: templates/fine-tune-llm_v2/training_configs/custom/mistralai/mistral-7b/lora/4xA10-512.yaml [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+# By default, we will keep the best checkpoint. You can change this to keep more checkpoints.
+num_checkpoints_to_keep: 1
+
+# Deepspeed configuration, you can provide your own deepspeed setup
+deepspeed:
+  config_path: deepspeed_configs/zero_3_offload_optim+param.json
+
+# Accelerator type, we value of 0.001 is not important, as long as it is
+# beteween 0 and 1. This ensures that accelerator type is used per trainer
+# worker.
```
Comment (SumanthRH) [inline templates/fine-tune-llm_v2/training_configs/custom/mistralai/mistral-7b/lora/4xA10-512.yaml:39]:
```suggestion
# Accelerator type, the value of 0.001 is not important, as long as it is
# between 0 and 1. This ensures that the given accelerator is available for each trainer
# worker.
```

File: templates/fine-tune-llm_v2/training_configs/custom/mistralai/mistral-7b/full/16xA10-512.yaml [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+# Rank-0 does fp32 aggregation, so we need 4x7B ~ 28GB of memory
+# Setting it to 35 GB to be safe
+# This should provision a large enough node to schedule rank-0
+trainer_resources:
+  memory: 37_580_963_840
+
+# Accelerator type, we value of 0.001 is not important, as long as it is
+# beteween 0 and 1. This ensures that accelerator type is used per trainer
+# worker.
```
Comment (SumanthRH) [inline templates/fine-tune-llm_v2/training_configs/custom/mistralai/mistral-7b/full/16xA10-512.yaml:45]:
```suggestion
# Accelerator type, the value of 0.001 is not important, as long as it is
# between 0 and 1. This ensures that the given accelerator is available for each trainer
# worker.
```

File: templates/fine-tune-llm_v2/training_configs/custom/mistralai/Mixtral-8x7B/lora/16xA10-4k.yaml [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+# By default, we will keep the best checkpoint. You can change this to keep more checkpoints.
+num_checkpoints_to_keep: 1
+
+# Deepspeed configuration, you can provide your own deepspeed setup
+deepspeed:
+  config_path: deepspeed_configs/zero_3.json
+
+# Accelerator type, we value of 0.001 is not important, as long as it is
+# beteween 0 and 1. This ensures that accelerator type is used per trainer
+# worker.
```
Comment (SumanthRH) [inline templates/fine-tune-llm_v2/training_configs/custom/mistralai/Mixtral-8x7B/lora/16xA10-4k.yaml:39]:
```suggestion
# Accelerator type, the value of 0.001 is not important, as long as it is
# between 0 and 1. This ensures that the given accelerator is available for each trainer
# worker.
```

File: templates/fine-tune-llm_v2/training_configs/custom/meta-llama/Meta-Llama-3-8B/lora/4xA10-512.yaml [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+# By default, we will keep the best checkpoint. You can change this to keep more checkpoints.
+num_checkpoints_to_keep: 1
+
+# Deepspeed configuration, you can provide your own deepspeed setup
+deepspeed:
+  config_path: deepspeed_configs/zero_3_offload_optim+param.json
+
+# Accelerator type, we value of 0.001 is not important, as long as it is
+# beteween 0 and 1. This ensures that accelerator type is used per trainer
+# worker.
```
Comment (SumanthRH) [inline templates/fine-tune-llm_v2/training_configs/custom/meta-llama/Meta-Llama-3-8B/lora/4xA10-512.yaml:39]:
```suggestion
# Accelerator type, the value of 0.001 is not important, as long as it is
# between 0 and 1. This ensures that the given accelerator is available for each trainer
# worker.
```

File: templates/fine-tune-llm_v2/training_configs/custom/meta-llama/Meta-Llama-3-8B/full/16xA10-512.yaml [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+# Rank-0 does fp32 aggregation, so we need 4x8B ~ 32GB of memory
+# Setting it to 35 GB to be safe
+# This should provision a large enough node to schedule rank-0
+trainer_resources:
+  memory: 37_580_963_840
+
+# Accelerator type, we value of 0.001 is not important, as long as it is
+# beteween 0 and 1. This ensures that accelerator type is used per trainer
+# worker.
```
Comment (SumanthRH) [inline templates/fine-tune-llm_v2/training_configs/custom/meta-llama/Meta-Llama-3-8B/full/16xA10-512.yaml:45]:
```suggestion
# Accelerator type, the value of 0.001 is not important, as long as it is
# between 0 and 1. This ensures that the given accelerator is available for each trainer
# worker.
```

File: templates/fine-tune-llm_v2/training_configs/custom/meta-llama/Meta-Llama-3-70B/full/16xA100-80G-4k.yaml [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+# Change this to the number of epochs that you want to train for
+num_epochs: 3
+
+# Change this to the batch size that you want to use
+train_batch_size_per_device: 2
+eval_batch_size_per_device: 2
+
+# Change this to the learning rate that you want to use
+learning_rate: 1e-4
```
Comment (SumanthRH) [inline templates/fine-tune-llm_v2/training_configs/custom/meta-llama/Meta-Llama-3-70B/full/16xA100-80G-4k.yaml:25]: Nit: Isn't this high for full param? Should be 5e-6

**Review by SumanthRH — COMMENTED** [review]
Thanks! I've left mostly nits. The main template readme looks great now! There are a couple configs that need `llmforge` changes, please see the specific comments below .

>  For some reason the runs on llama-3-70B were insanely slow on g5 (even its lora) I think we need to do some profiling and optimization.

I'm a bit surprised by the slowdown, need to check. For lora, it looks like it was single node with cpu offloading, so it could just be that. I've noticed that even Llama-3-8B needs cpu offloading with 8xA10s for some context lengths, so heavy offloading might be one reason. 


For full-param, we need multi-node and frankly if we made the settings for A100s , we could ahve used heirarchical partitioning: https://www.microsoft.com/en-us/research/blog/deepspeed-zero-a-leap-in-speed-for-llm-and-chat-model-training-with-4x-less-communication/. But what we have with A10s is okay I guess.

File: templates/fine-tune-llm_v2/README.ipynb [hunk 16]
```diff
@@ -6,142 +6,341 @@
...
-   "outputs": [],
    "source": [
-    "# [Optional] You can set the WandB API key to track model performance\n",
-    "# import os\n",
-    "# os.environ[\"WANDB_API_KEY\"]=\"YOUR_WANDB_API_KEY\"\n",
+    "### Models Supported in Default Mode\n",
+    "\n",
+    "Default mode supports a select list of models, with a fixed cluster type of 8xA100-80G. For each model we only support context lengths of 512 up to Max. context length in increments of 2x (i.e. 512, 1024, ...). Here are the supported models and their configurations:\n",
+    "\n",
+    "Model family | model_id(s) | Max. context lengths |\n",
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply templates/fine-tune-llm_v2/README.ipynb:84]: it renders ok on vscode. I have added what I think was the main issue. But regardless this readme won't stay her for long.

File: templates/fine-tune-llm_v2/README.ipynb (hunk 2 above)
Comment (kouroshHakha) [reply templates/fine-tune-llm_v2/README.ipynb:140]: It's ok I didn't mean for this to be complete. It's just meant to be part of an example.

File: templates/fine-tune-llm_v2/README.ipynb (hunk 5 above)
Comment (kouroshHakha) [reply templates/fine-tune-llm_v2/README.ipynb:151]: yeah sure but this is meant for sinking in the idea not the semantics. We can fix things better once people start facing issues and report problems back. I think it's ok.

**Review by SumanthRH — APPROVED** [review]
LGTM ! Thanks!

**Review by akshay-anyscale — APPROVED** [review]

**Pull request merged**